### PR TITLE
[WIN32SS][NTGDI] Text/Font Rotation/Transformation with optimization

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-11848](https://jira.reactos.org/browse/CORE-11848), [CORE-16020](https://jira.reactos.org/browse/CORE-16020)
- Enable text/font rotation/transformation by `lfEscapement` and `WorldTransform`.
- Split `IntExtTextOutW` function to `IntExtTextOutZeroAngleW` and `IntExtTextOutAnyAngleW` functions.
- Add `IntNormalizeAngle` and `IntScaleMatrix` functions.
- Delete `ScaleLong` function.
- add `const` to some parameters of some `FLOATOBJ_*` functions.

## TODOs
- [ ] Optimize more.
- [ ] Do tests.

### CODE REVIEW

**Comment by katahiromz** [comment]
COMPARE:
https://github.com/reactos/reactos/blob/master/win32ss/gdi/ntgdi/freetype.c
https://github.com/reactos/reactos/blob/86046a9bf4af57551308331631447c7ccfa87fe9/win32ss/gdi/ntgdi/freetype.c

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -6216,15 +6189,789 @@ IntExtTextOutW(
...
+    if (!TextObj)
+        return FALSE;
+
+    FontObj = TextObj->Font;
+    ASSERT(FontObj);
+
+    FontGDI = ObjToGDI(FontObj, FONT);
+    ASSERT(FontGDI);
+
+    plf = &TextObj->logfont.elfEnumLogfontEx.elfLogFont;
```
Comment (Doug-Lyons) [inline win32ss/gdi/ntgdi/freetype.c]: if ((plf->lfEscapement != 0) || (pdcattr->iGraphicsMode != GM_ADVANCED))  /* Bypass most tests */
{

File: win32ss/gdi/ntgdi/freetype.c [hunk 2]
```diff
@@ -6216,15 +6189,789 @@ IntExtTextOutW(
...
+        if (pdcattr->iGraphicsMode == GM_ADVANCED)
+        {
+            pmxWorldToDevice = DC_pmxWorldToDevice(dc);
+            bZeroAngle = !!(pmxWorldToDevice->flAccel & XFORM_SCALE);
+        }
+        else
+        {
+            bZeroAngle = TRUE;
+        }
+    }
```
Comment (Doug-Lyons) [inline win32ss/gdi/ntgdi/freetype.c]: } /* End of tests bypass */

File: win32ss/gdi/ntgdi/freetype.c [hunk 3]
```diff
@@ -6216,15 +6189,789 @@ IntExtTextOutW(
...
+    IN DWORD dwCodePage)
+{
+    PDC_ATTR pdcattr;
+    FONTOBJ *FontObj;
+    PFONTGDI FontGDI;
+    LOGFONTW *plf;
+    LONG lfEscapement;
+    PTEXTOBJ TextObj;
+    PMATRIX pmxWorldToDevice;
+    BOOL bResult = FALSE, bZeroAngle;
```
Comment (Doug-Lyons) [inline win32ss/gdi/ntgdi/freetype.c]: BOOL bResult = FALSE, bZeroAngle = TRUE;  /* Use bZeroAngle default as TRUE */

**Comment by JoachimHenze** [comment]
Thx Katayama for the improved version.
I propose that the tests should also cover the real-world-testcase of RawTherapy from CORE-16020 again. I propose we experiment with that in a private build before merging the PR to master.

**Comment by katahiromz** [comment]
wrong
Wrong.

**Comment by katahiromz** [comment]
fixing
OK. Let's enter the testing stage...

**Comment by katahiromz** [comment]
success
Success! @JoachimHenze Let's go!

File: win32ss/gdi/ntgdi/freetype.c [hunk 4]
```diff
@@ -6216,15 +6188,787 @@ IntExtTextOutW(
...
+                                         fuOptions,
+                                         lprc,
+                                         String, Count,
+                                         Dx,
+                                         dwCodePage,
+                                         TextObj);
+    }
+
+    if (TextObj)
+        TEXTOBJ_UnlockText(TextObj);
```
Comment (learn-more) [inline win32ss/gdi/ntgdi/freetype.c]: `TextObj` is always set here

**Comment by katahiromz** [comment]
wrong
Wrong in RawTherapee.

File: win32ss/gdi/ntgdi/freetype.c (hunk 4 above)
Comment (learn-more) [reply win32ss/gdi/ntgdi/freetype.c]: (so no need for the NULL-check)

**Comment by katahiromz** [comment]
Gotcha!
gotcha

Adding clipping optimization and cache optimization...

File: win32ss/gdi/ntgdi/freetype.c [hunk 5]
```diff
@@ -705,14 +705,28 @@ LONG FASTCALL IntWidthMatrix(FT_Face face, FT_Matrix *pmat, LONG lfWidth)
...
+FT_Fixed FASTCALL FT_FixedFromFLOATOBJ(const FLOATOBJ *pef)
+{
+    FLOATOBJ efTemp;
+    if (FLOATOBJ_Equal1(pef))
+    {
+        return INT_TO_FIXED(1);
+    }
+
+    efTemp = *pef;
+    FLOATOBJ_MulLong(&efTemp, INT_TO_FIXED(1));
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/freetype.c]: i don't understand that, can you explain to me?

Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: `INT_TO_FIXED` is defined in `$(FREETYPE)\include\freetype\internal\ftcalc.h` as follows:
```c
#define INT_TO_FIXED( x )      ( (FT_Long)(x) * 65536 )  /* << 16 */
```
Thus this macro makes an integer value a 16.16 fixed point (`FT_Fixed`). `FIXED` means `FT_Fixed` in FreeType. This code is multiplication of a `FLOATOBJ` value and a `FT_Fixed` value. You can see `FT_Matrix` in `$(FREETYPE)\include\freetype\fttypes.h`, that has its `FT_Fixed` members. `FLOATOBJ_GetLong` will return the integral part of a value multiplied by `65536`.

**Comment by katahiromz** [comment]
Do you have any ideas how to optimize more?

File: win32ss/gdi/ntgdi/freetype.c [hunk 6]
```diff
@@ -6216,15 +6202,786 @@ IntExtTextOutW(
...
+    {
+        bZeroAngle = FALSE;
+    }
+    else if (pdcattr->iGraphicsMode == GM_ADVANCED)
+    {
+        pmxWorldToDevice = DC_pmxWorldToDevice(dc);
+
+        if (!(pmxWorldToDevice->flAccel & XFORM_SCALE) ||
+            FLOATOBJ_GetFix(&pmxWorldToDevice->efM11) < 0 ||
+            FLOATOBJ_GetFix(&pmxWorldToDevice->efM22) < 0)
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/freetype.c]: To be fixed.

**Comment by katahiromz** [comment]
image
There are 5803 calls of `SetWorldTransform`. We have to optimize `SetWorldTransform`.

**Comment by katahiromz** [comment]
sluggish
RawTherapee is still 'sluggish'.

**Comment by katahiromz** [comment]
Temporarily closed.

**Comment by katahiromz** [comment]
Reopened.

**Comment by katahiromz** [comment]
fixing-sluggish
Sluggishness is a little reduced.

**Comment by katahiromz** [comment]
failed
Buggy. Temporarily closed.

**Comment by HBelusca** [comment]
So.... what happened with this PR?

**Comment by katahiromz** [comment]
I retired.

**Comment by JoachimHenze** [comment]
In case it will be reopened: we need to retest CORE-16020 and CORE-16177 before committing.

**Pull request closed without merging**